### PR TITLE
Make EndpointSelector.select() aware of current request

### LINKED ISSUE #766: Make EndpointSelector.select() aware of current request
Some load-balancing mechanism uses a property of the current request as its input, e.g. the current user's ID and the current request's certain parameter. However, it's not possible to implement such a mechanism with our endpoint API at the moment because `EndpointSelector.select()` does not accept any parameters. I propose to add `ClientRequestContext` to the parameter list of:

- `EndpointSelector.select()`
- `EndpointGroupRegistry.selectNode()`
- `Endpoint.resolve()`

so that a user can inject some information which could be used for load-balancing into `ClientRequestContext` attribute map.

### PR DESCRIPTION
**Pull request by imasahiro**
Fixes #766

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/client/endpoint/EndpointGroupRegistry.java [hunk 1]
```diff
@@ -103,16 +104,16 @@ public static EndpointGroup get(String groupName) {
 
     /**
      * Selects an {@link Endpoint} from the {@link EndpointGroup} associated with the specified
-     * case-insensitive {@code groupName}.
+     * {@link ClientRequestContext} and the specified case-insensitive {@code groupName}.
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/endpoint/EndpointGroupRegistry.java]: Can remove `the specified`?

**Comment by trustin** [comment]
@imasahiro We can merge this once @minwoox's comment is addressed.

File: core/src/main/java/com/linecorp/armeria/client/endpoint/EndpointGroupRegistry.java (hunk 1 above)
Comment (imasahiro) [reply core/src/main/java/com/linecorp/armeria/client/endpoint/EndpointGroupRegistry.java]: Removed

**Comment by imasahiro** [comment]
@anuraaga Thanks for suggestions. I didn't dig it yet but maybe it can be possible if client decorator attaches user-id through e.g. request header? Let me think a bit in next PR.

**Comment by anuraaga** [comment]
Cool - yeah I think it might not affect this API. Using a decorator to move a header into the context seems very heavyweight, but I think propagating through a threadlocal, something like this might be pretty reasonable

```java
try (withContextCustomizer(ctx -> ctx.attrs().set(USER_ID, userId)) {
  client.doRpc(request);
}
```

**Comment by trustin** [comment]
Thanks, @imasahiro and reviewers!

**Pull request merged**